### ISSUE TITLE
TlsServerAuthorizationCheckConfig may cause SIGSEGV

### ISSUE THREAD

**Issue by exiledkingcc** [grpc/grpc#24684]
this is the code which causes SIGSEGV
```cpp
static std::shared_ptr<Channel> makeChannel() {
    auto tls_key_cfg_sp = std::make_shared<TlsKeyMaterialsConfig>();
    // set key and certs...
    auto tls_sac_sp = std::make_shared<TlsServerAuthorizationCheckConfig>(
        std::make_shared<MyServerAuthorizationCheck());
    auto opts = grpc::experimental::TlsCredentialsOptions{
        grpc_tls_server_verification_option::GRPC_TLS_SKIP_ALL_SERVER_VERIFICATION,
        tls_key_cfg_sp, nullptr, tls_sac_sp
    };
    auto credentials = grpc::experimental::TlsCredentials(opts);
    ChannelArguments arguments;
    arguments.SetSslTargetNameOverride("...");
    return grpc::CreateCustomChannel("...", credentials, arguments);
}

class MyClient {
    explicit TransMsgClient():stub_(MyMsgSrv::NewStub(makeChannel())){}
   // some methods with stub_
private:
    std::unique_ptr<MyMsgSrv::Stub> stub_;
};

auto cl = MyClient();
// do somthing with cl
```

this is how I fix it: 
make  std::shared_ptr<TlsServerAuthorizationCheckConfig> has longer lifetime than Channel
```cpp
static std::shared_ptr<TlsServerAuthorizationCheckConfig> tls_sac_sp;
// init then use it in makeChannel
```

i do NOT think it's nesscessary to keep a shared_ptr of TlsServerAuthorizationCheckConfig.
it's a bug.
std::shared_ptr should NOT be moved.

**Comment by ZhenLian** [comment]
@exiledkingcc Thanks for reporting this bug! I will be working on this in the next couple of days.
For what it's worth, we are in the middle of re-design and re-implement the whole TlsCredentials API. In the new API, `TlsKeyMaterialsConfig` is removed, and `TlsServerAuthorizationCheckConfig` might change slightly. You can find more information here: https://github.com/grpc/proposal/pull/205

**Comment by peterhurley-meraki** [comment]
This is still a problem with the new API in 1.38.x

**Comment by peterhurley-meraki** [comment]
The fundamental problem is that grpc_tls_server_authorization_check_config has a context_ pointer to its "owning" TlsServerAuthorizationCheckConfig which is stale when the TlsServerAuthorizationCheckConfig object is freed (eg., when it goes out of scope).

The access violation happens when that context_ is dereferenced in TlsServerAuthorizationCheckConfigCSchedule  [called from TlsChannelSecurityConnector::check_peer -> grpc_tls_server_authorization_check_config::Schedule()].

Once a channel is created, nothing preserves the c++ object chain-of-ownership from TlsCredentials -> TlsCredentialsOptions -> TlsServerAuthorizationCheckConfig.

**Comment by peterhurley-meraki** [comment]
This issue is actually the same issue that was reported here https://github.com/grpc/grpc/issues/22287 but never responded to.

**Comment by ZhenLian** [comment]
Yeah, we are aware of this issue and wanted to get rid of it as soon as possible.
The whole custom verification API is under re-design and re-implementation in https://github.com/grpc/grpc/pull/25631. Once that is merged, all API related to `TlsServerAuthorizationCheckConfig` and `TlsServerAuthorizationCheckConfigCSchedule` will be replaced with a new verifier-based API. 
@yihuazhang FYI.

**Comment by erm-g** [comment]
@exiledkingcc @peterhurley-meraki I see that [25631](https://github.com/grpc/grpc/pull/25631) was merged a while ago, is the issue fixed?

**Comment by matthewstevenson88** [comment]
Closing since the code in question has been deleted.
